### ISSUE TITLE
global rule tracking (stylesheet.replace)

### ISSUE THREAD

**Issue by jschr** [threepointone/glamor#222]
I'm using `css.global` to set the `html` font-size for rems based on `props.baseFontSize`. Everything works when `props.baseFontSize` is changed but for example, if it were to change from 16 to 22 and then back to 16 it will not apply the styles of the last prop of 16. 

You can see the issue here: http://theme-provider-basic-example.surge.sh. Change the font size to 22 and then back to 16. You'll see that additional rules are added to the stylesheet but the `html` element's styles aren't updated to previous values.

Is this expected? 

I'm using glamor 2, I haven't tried v3 yet.

**Comment by ChristopherBiscardi** [comment]
This is the expected behavior. The intent of `.global` (and `.insert`) is as an escape hatch to append a single raw rule to the stylesheet. It does not allow for the updating of these rules.

The behavior in the example (thank you for taking the time to set that up!) will append any new rule to the stylesheet and because source ordering, the last rule appended takes precedence. The rules that have been inserted previously are stored in a cache and thus will not be re-inserted when switching back to a previous font-size.

Is there a specific use case you are trying to implement?

**Comment by jschr** [comment]
Thanks for the explanation, the behaviour makes sense knowing that. 

My use case is not typical, I'm building a theme generator app where the backend will present a UI for previewing simple changes like base font size, primary colour, etc... and output a configured application. 

I built a [small lib](https://github.com/jschr/theme-provider) to help with theming components. It accounts for theming based on props, which the backend relies on to preview theme changes before publishing. So far I have no had any real-world use cases for theming based on props though... so I wouldn't expect most of glamor's users to run into this issue.

Is it possible to clear all of the styles generated by glamor? A workaround for me would be to clear all the styles before re-creating any new css rules via prop changes.

**Comment by jschr** [comment]
Knowing a bit how glamor works now, am I correct in thinking that I should not be using `css({ ... })` in the render function because this will cause new rules to be generated every render? Since I'm creating a new object each time.

**Comment by donaldpipowitch** [comment]
Yes, if you don't need to change your style based on new props, it is better to call `css()` just once outside of `render`.

**Comment by ChristopherBiscardi** [comment]
There is also [glamor babel plugin](https://github.com/threepointone/glamor/blob/f4fce31163d0002331819fdb66a77e998ba6825d/docs/weakmaps.md#so-how-can-we-use-it-with-glamor) which gets a couple mentions in the docs but isn't really covered enough.

For a workaround, [flush](https://github.com/threepointone/glamor/blob/518e63ca1a12476f40dcb665d32b3b4083e6dde2/src/sheet.js#L191) might be useful to you.

**Comment by jschr** [comment]
Thanks, flush is what I was looking for! It should provide an acceptable workaround. 

You're welcome to close this issue if glamor doesn't intend to support this use case.

**Comment by ChristopherBiscardi** [comment]
I think we're pretty close to having something like this given that [stylesheet.replace is scaffolded out](https://github.com/threepointone/glamor/blob/518e63ca1a12476f40dcb665d32b3b4083e6dde2/src/sheet.js#L170). That combined with `stylesheet.insert` [returning the internal counter](https://github.com/threepointone/glamor/blob/518e63ca1a12476f40dcb665d32b3b4083e6dde2/src/sheet.js#L151) could mean that a user would be able to set up their own state to hold the index to operate on via replace later. Does that sound like it would fit your use case? If so I'll swap the title of the issue to reflect that to re-orient the discussion.

in pseudo-code:

```jsx
componentDidMount() {
  this.setState({
    htmlIndex: css.insert('html', `font-size: ${this.state.fontSize}`)
  })
}
onChangeFontSize(size) {
  stylesheet.replace(this.state.htmlIndex, `font-size: ${this.state.fontSize}`);
}
render() {
}
```

I don't know what @threepointone was envisioning for `replace` in v3 though.

**Comment by jschr** [comment]
Yea that sounds like it would be a useful feature to have beyond `css.global`, welcome to change the title of the issue.

Another use case I've had for the being able to replace the stylesheet:

```js
function lineClamp(num) {
  return {
    display: '-webkit-box',
    WebkitLineClamp: num,
    WebkitBoxOrient: 'vertical',
    overflow: 'hidden'
  }
}

function TextClamp({ text, clamp }) {
  return (
    <span { ...css(lineClamp(clamp)) }>
      { text }
    </span>
  )
}
```

Since it's recreating the object every time I wouldn't run into the original rule ordering issue but being able to replace the stylesheet to prevent unnecessary rules from being generated on every render would be nice to have.

**Comment by otbe** [comment]
I think your last example should work as expected without generating new rules on every render, because glamor will evaluate your style object and calculate the hash. If the hash is already present it won't add another rule.

**Comment by jschr** [comment]
@otbe you're right, I updated my example page (http://theme-provider-basic-example.surge.sh). 

If you change line clamp you'll see that a rule is generated for every value but only once. Although now I'm not sure why I don't see the ordering problem like I do with `css.global` (the original issue)....

**Comment by ChristopherBiscardi** [comment]
@jschr The original issue doesn't occur because the target element is a classname (`data-css-1i87vau`) whose hash is different and applied to the `span` at the exclusion of the other classnames. Therefore there are no conflicts that would fall back on source ordering:

<img width="226" alt="screen shot 2017-03-13 at 10 40 04 am" src="https://cloud.githubusercontent.com/assets/551247/23867599/a2370020-07d9-11e7-9e6b-cc514c56c306.png">
<img width="223" alt="screen shot 2017-03-13 at 10 40 19 am" src="https://cloud.githubusercontent.com/assets/551247/23867602/a48e0be8-07d9-11e7-9146-e1c613c706c9.png">

You can see that the target selectors for the original issue are exactly the same, and therefore fall back to source ordering

<img width="568" alt="screen shot 2017-03-13 at 10 45 18 am" src="https://cloud.githubusercontent.com/assets/551247/23867772/3c6aad5e-07da-11e7-81e5-15dab5ea07ba.png">

**Comment by jschr** [comment]
thanks for the explanation, I was confused at first why I didn't see the issue but then realized that it's the global selectors that only have the problem 👍

**Comment by otbe** [comment]
Our rule of thumb is: avoid global CSS at all :D

**Comment by jschr** [comment]
Definitely, the only cases where I've *needed* to use global css is for setting the root font-size on `html` for rems and setting `height: 100%` on `html, body, #react-root`.

**Comment by jschr** [comment]
With regards to the global issue, as an alternative to replace -- curious if it would be possible for glamor to add `data-css-*` to the `html` element and generate a rule for the classname instead of adding adding a rule for `html`?

**Comment by donaldpipowitch** [comment]
Maybe you can do that manually? Just get `data-css-` from glamor and add it as a CSS class with classic DOM APIs?

**Comment by jschr** [comment]
Thats a great point, I just added that workaround to the test page it works!

Here's the code:

```js
let previousRule

// called on props change
function createStyles(theme) {
  const htmlCss = css({
    fontFamily: theme.fontFamily,
    fontSize: `${theme.baseFontSize}px`
  })

  const rule = Object.keys(htmlCss)[0]

  document.documentElement.removeAttribute(previousRule)
  document.documentElement.setAttribute(rule, htmlCss[rule])

  previousRule = rule
}
```

**Comment by lopis** [comment]
Another more common use case: when a user opens a menu (e.g. through a hamburger) or a modal which contains scrollable content, it may interfere with the page scroll. I was trying to use global css to toggle "overflow" in the body.

**Comment by threepointone** [comment]
appears the workaround helps, closing this issue.
